Blog post about e-voting schemes (#63)
lunatech-labs/lunatech-blog: posts/2022-10-06-experimenting-with-voting-schemes.adoc

= Introduction to e-voting schemes
agoetschm
v1.0, 2022-10-06
:title: Introduction to e-voting schemes
:lang: en
:tags: [voting, crypto, zero-knowledge, en]

Given the availability of technology in our society as well as the need for tools supporting democratic decision making, it is not surprising that e-voting is a growing research field.

> But unlike suggested by their name, e-voting schemes are not only digitalized version of traditional voting systems where paper ballots are counted by hand.

Reviewer: ```suggestion
But unlike suggested by their name, e-voting schemes are not only digitalized versions of traditional voting systems where paper ballots are counted by hand.
```

They go a step further by leveraging cryptographic methods to provide privacy and verifiability guarantees. 
In particular, trust is shifted from the system to independent verifiers who can check that everything went well.

The aim of this post is to gain an understanding of how this type of systems works with a concrete example. We will go over some background information on voting schemes, give a description of the Helios 2.0 scheme, look into a possible attack and finish with a few pointers about key parts of a simplified implementation.

== Types of voting schemes

> In traditional voting systems, each voter puts a piece of paper carrying her vote in a box. Some trusted administrator then opens the box, counts the votes and publishes the result.

Reviewer: Traditional conundrum with choosing sexes for a subject. Solution is to either utilise her/his or use the proposed change.

```suggestion
In traditional voting systems, each voter puts a piece of paper carrying their vote in a box. Some trusted administrator then opens the box, counts the votes and publishes the result.
```

Author: I will leave it like this, a little feminism shouldn't hurt here :)

The class of schemes we are talking about follow the same pattern, except that instead of casting a piece of paper in a box, the vote is encrypted before being shared on a public bulletin board, i.e. a medium where anyone can read from. 
This allows observers to check some properties of the ballots without seeing their content since they are encrypted.
The ballots can then be tallied by the organizer in possession of the necessary decryption material.

> .Voting process with some details we will get into later.

Reviewer: Is this a caption on the diagram that follows?

Suggestion: I would replace the ascii art based diagrams with 'real' drawings. Do you have access to a drawing tool such as Omnigraffle? If not, I think we should still convert the drawings and see who could make the drawings for you (there aren't that many in the article anyway).

Author: yes it's the caption

you're right proper diagrams should make it a bit cleaner

 ...    -----------+->         ...         ---+-------+
                   |                          |       |
                   |  +--------------------+  |       |    +------------------+   reveal
vote n  -----------+->| encrypted ballot n +--+-------+--->| encrypted result +------------> result
                   |  +--------------------+  |            +------------------+
                   |                          |
                   +--------------------------+
----

Voting schemes of this type aim at shifting the trust of the voters away from the system to some data published during the process.
This means that instead relying on the system as a black-box, voters can be convinced of the integrity of the result by examining the public bulleting board.
In parallel of being verifiable, these voting schemes naturally try to keep the content of each ballot hidden to preserve the privacy of the voters.
We will see how this can be achieved in the next section.

In order to limit the scope of this post, we do not worry about the eligibility of the voters. In other words, we assume that another system checks whether a vote comes from a legitimate voter or not.

> But this does not prevent legitimate voters from colluding and potentially learning unexpected information, as we will see <<attack, later>>. 

Reviewer: ```suggestion
But this does not prevent legitimate voters from colluding and potentially learning information that should never be revealed to start with, as we will see <<attack, later>>. 
```

....
+------+     +------------------+     +------------------+     +--------+
| vote +---->| encrypted ballot +---->| encrypted result +---->| result |
+------+  |  +------------------+     +------------------+  |  +--------+
          |                  |           ^                  |
          |                  |           |                  |
          v                  +-----------+                  v
        proof                reproducible                 proof

          |                        |                        |
          +------------------------+------------------------+
                         chain of verifiability
....

We will now describe how the voting scheme is used and built before coming back to the necessary building blocks, namely the encryption scheme and the zero-knowledge proofs.

=== Overview of the scheme

In this part we describe the participants of the voting process and which algorithms they execute.

==== Participants

> As mentioned earlier, three kinds of participants have a role in this voting process. First comes the __organizer__, who takes care of generating the necessary encryption/decryption keys at the beginning and reveals the voting result at the end. The organizer has to be trusted to not share his part of the key. Second are the _voters_, who make a choice regarding their vote and cast an encrypted ballot. Lastly come the _verifiers_, who check the steps of the verifiability chain to make sure of the integrity of the process. Anyone, voter or external, can act as a verifier.

Reviewer: ```suggestion
As mentioned earlier, three kinds of participants have a role in this voting process. First comes the __organizer__, who takes care of generating the necessary encryption/decryption keys at the beginning and reveals the voting result at the end. The organizer has to be trusted to not share his part of the key. Second are the _voters_, who, as the their name suggests, vote and cast an encrypted ballot. Lastly come the _verifiers_, who check the steps of the verifiability chain to make sure of the integrity of the process. _Anyone_, voter or external, can act as a verifier.
```

`+vote+`:: Each voter uses this function to construct a ballot from her vote. This implies the encryption of the vote using the public key as well as the generation of some zero-knowledge proofs attesting that the vote is valid. 
`+cast+`:: After building the ballot, each voter sends it to the public bulletin board, where it is visible to everyone.
`+tally+`:: This step can be executed by anyone thanks to the type of encryption used. The ballots stay encrypted while they are combined to form the encrypted result. More details in the <<encryption, next part>>.

> `+reveal+`:: In the final step, the organizer uses her secret key to decrypt the encrypted result and reveals the voting result. A corresponding proof is also generated to attest that the result is indeed the decryption of the tally result.

Reviewer: ```suggestion
`+reveal+`:: In the final step, the organizer uses their secret key to decrypt the encrypted result and reveals the voting result. A corresponding proof is also generated to attest that the result is indeed the decryption of the tally result.
```

                              |                         |                          | result <- reveal(tally(ballots), secret key)
                              |                         |          result          |
                              |                         | <----------------------- |
                              |                         |                          |
                              |                         |                          |
....

We will now discuss in more details how the encryption and the proofs of knowledge work in Helios.

=== Encryption [[encryption]]

In general, encryption is a method to hide information and to be able to transmit it over an exposed channel. 
Typically, a function called `encrypt` converts a plaintext message `m` into a ciphertext `c` which does not reveal anything about `m`. 
Another function `decrypt` transforms `c` back into the recovered plaintext `m'`, which should be identical to `m`.

....
+---+  encrypt  +---+  decrypt  +---+
| m +---------->| c +---------->| m'|
+---+     ^     +---+     ^     +---+
          |               |
          |               |
     public key      secret key
....

> In order execute those functions, another piece of information called __key__ is needed.

Reviewer: ```suggestion
In to order execute those functions, another piece of information called __key__ is needed.
```

.Details of Alice's and Bob's encrypted ballots
[%collapsible]
====

* Alice wants to vote for the first candidate and Bob for the second: `v~a~ = (1, 0, 0)` and `v~b~ = (0, 1, 0)`
* let `q = 53`, `p = 2q + 1 = 107` and `g = 75`: `+g+` generates a subgroup of `Z/pZ` of order `q`
* (note that we have not mentioned `q` earlier for simplicity, but it does not change the rest of this example)
* let the key pair be `x = 18` and `h = g^x^ = 35 mod p`
* the encrypted ballots can be calculated as follows: 
** choose a random value, say `r = 2`
** encryption of the first element of Alice's vote: `b~a1~ = (g^r^, h^r^ * g^m^) = (61, 48 * 75) = (61, 69)`
** note that the vote is projected in `Z/pZ` before being encrypted with `g^m^`
** similarly for the first element of Bob's vote with `r = 3`: `b~b1~ = (81, 75 * 1) = (81, 75)`
* the encrypted sum for the first candidate is: `c~1~ = b~a1~ * b~b1~ = (61 * 81, 69 * 75) = (19, 39)`
* and once decrypted: `s~1~ = 39 * (19^x^)^-1^ = 39 * 69^-1^ = 39 * 76 = 75 = g^1^`

> * the sum of the votes for the first candidate is indeed `1`

Reviewer: I think that this section could benefit from some extra clarifications.

For example, what does "the vote is projected `Z/pZ`" mean. In the same sentence what is `m` in `g^m`?

In the sentence "and once decrypted: ..." what is done to get from `19^x^-1` to `69^-1`?

I did the calculation in the Scala REPL. I wonder if putting a small VSC/Metals worksheet in the project wouldn't be a good idea...

Author: good points
I added a link to a scastie snippet, this way there is no need to clone the repo

.Schorr's protocol
----
+-----------------+                +-----------------+
|     Prover      |                |    Verifier     |
|                 |                |                 |
| choose random r |                |                 |
|     u = g^r     |       u        |                 |
|                 | -------------> |                 |
|                 |                | choose random c |
|                 |       c        |                 |
|                 | <------------- |                 |
|  z = r + x * c  |                |                 |
|                 |       z        |                 |
|                 | -------------> |                 |
|                 |                |    check if     |
|                 |                |  g^z = u * h^c  |

> +-----------------+                +-----------------+

Reviewer: Maybe it would be good to point out what is private and public information in this scheme?

----

> P first choose a random value `r` and commits to it by sending `u = g^r^`. Next, V choose a random challenge `c` and sends it back. Lastly, P sends `z` such that `g^z^ = u * h^c^` and the only value satisfying this equality is `z = r + x * c`. She can only calculate `z` knowing `x`, thus V deduces that with a very high probability P indeed knows `x` such that `h = g^x^`.

Reviewer: Eliminate the need for using he/she. I hope my replacement with `V` is correct.

```suggestion
P first choose a random value `r` and commits to it by sending `u = g^r^`. Next, V choose a random challenge `c` and sends it back. Lastly, P sends `z` such that `g^z^ = u * h^c^` and the only value satisfying this equality is `z = r + x * c`. V can only calculate `z` knowing `x`, thus V deduces that with a very high probability P indeed knows `x` such that `h = g^x^`.
```

  val exponent: F
  val generator: Gen
----
The https://github.com/agoetschm/voting-helios-scala/tree/develop/src/main/scala/algebra/Domain.scala[Domain] trait englobes a discrete group `base`, which is used later as the ElGamal public key space, with a finite field `exponent`, corresponding to the private key space. The `generator` projects elements of the `exponent` field on a subgroup of the `base` group. Note that `Domain` is "data type agnostic": the type parameter `Z` can be any type representing an integer - such as `Int`, `Long` or `BigInt` - in order to handle different sizes of numbers.

On top of this is built the https://github.com/agoetschm/voting-helios-scala/tree/develop/src/main/scala/crypto/[crypto] package. It contains an implementation of the https://github.com/agoetschm/voting-helios-scala/tree/develop/src/main/scala/crypto/ElGamal.scala[ElGamal encryption scheme] and the https://github.com/agoetschm/voting-helios-scala/tree/develop/src/main/scala/crypto/ProofOfKnowledge.scala[proofs of knowledge] needed for Helios.
[source,scala]
----
trait EncryptionScheme[PT, CT, PK, SK, D, R]:
  val gen: (D, () => R) => (PK, SK)
  val enc: (PT, PK, () => R) => CT

>   val dec: (CT, SK) => PT

Reviewer: Suggestion; I would either add a legend describing the different type parameters or, alternatively swap the abbreviations for descriptive names. (`PT` -> `Plaintext`, ... `D` -> ?, `R` -> ?)
